File: 回測模組/analytics/trade_details.py
"""
交易詳情處理模組
負責收集和處理詳細的交易記錄
"""
from typing import Dict, List
from strategy_modules.position_manager import TradeRecord


class TradeDetailsProcessor:
    """交易詳情處理器"""

    def collect_trade_details(self, trades: List[TradeRecord], stock_id: str, date: str) -> List[Dict]:
        """
        收集單一股票的詳細交易記錄

        Args:
            trades: 交易記錄列表
            stock_id: 股票代碼
            date: 日期

        Returns:
            詳細交易記錄列表
        """
        trade_details = []

        for i, trade in enumerate(trades, 1):
            # 計算實際出場價格（加權平均）
            exit_price = self.calculate_actual_exit_price(trade)
            pnl = (exit_price - trade.entry_price) * 1000  # 1張股票
            pnl_percent = ((exit_price - trade.entry_price) / trade.entry_price) * 100

            # 處理移動停利出場（可能有多次）
            if trade.trailing_exit_details:
                self._add_trailing_exit_details(trade_details, trade, i, stock_id, exit_price, pnl, pnl_percent)

            # 處理兩階段出場模式
            elif trade.partial_exit_time or trade.final_exit_time:
                self._add_two_stage_exit_details(trade_details, trade, i, stock_id, exit_price, pnl, pnl_percent)

        return trade_details

    def calculate_actual_exit_price(self, trade: TradeRecord) -> float:
        """
        計算實際出場價格（加權平均）

        Args:
            trade: 交易記錄

        Returns:
            加權平均出場價格
        """
        if trade.trailing_exit_details:
            # 移動停利出場：計算加權平均價
            total_ratio = 0.0
            weighted_price = 0.0

            for exit_detail in trade.trailing_exit_details:
                weighted_price += exit_detail['price'] * exit_detail['ratio']
                total_ratio += exit_detail['ratio']

            # 如果還有剩餘部位（進場價保護或收盤）
            if trade.final_exit_price and total_ratio < 1.0:
                remaining_ratio = 1.0 - total_ratio
                weighted_price += trade.final_exit_price * remaining_ratio
                total_ratio = 1.0

            return weighted_price / total_ratio if total_ratio > 0 else trade.entry_price

        elif trade.partial_exit_price and trade.final_exit_price:
            # 兩階段出場
            if trade.reentry_price:
                # 有回補：50%在partial_exit，50%在final_exit
                return (trade.partial_exit_price * 0.5 + trade.final_exit_price * 0.5)
            else:
                # 無回補：50%在partial_exit，50%在final_exit
                return (trade.partial_exit_price * 0.5 + trade.final_exit_price * 0.5)

        elif trade.partial_exit_price:
            # 只有部分出場
            return trade.partial_exit_price

        elif trade.final_exit_price:
            # 只有最終出場
            return trade.final_exit_price

        else:
            # 沒有出場（理論上不應該發生）
            return trade.entry_price

    def _add_trailing_exit_details(self, trade_details: List[Dict], trade: TradeRecord, trade_num: int,
                                    stock_id: str, exit_price: float, pnl: float, pnl_percent: float) -> None:
        """添加移動停利出場詳情"""
        # 每次移動停利出場
        for j, exit_detail in enumerate(trade.trailing_exit_details):
            detail = {
                '股票代碼': stock_id,
                '交易編號': trade_num,
                '進場時間': trade.entry_time,
                '進場價格': trade.entry_price,
                '進場Ratio': trade.entry_ratio,
                '進場外盤3秒(M)': trade.entry_outside_volume_3s / 1000000 if trade.entry_outside_volume_3s else 0,
                '出場批次': j + 1,
                '出場類型': f"移動停利_{exit_detail['level']}",
                '出場時間': exit_detail['time'],
                '出場價格': exit_detail['price'],
                '出場比例': exit_detail['ratio'] * 100,
                '出場原因': f"跌破{exit_detail['level']}低點",
                '實際出場均價': exit_price,
                '總損益金額': pnl,
                '總損益百分比': pnl_percent
            }
            trade_details.append(detail)

        # 如果有進場價保護或收盤出場（剩餘部位）
        if trade.final_exit_time and trade.final_exit_reason:
            remaining_ratio = 1.0 - sum(exit['ratio'] for exit in trade.trailing_exit_details)
            if remaining_ratio > 0:
                detail = {
                    '股票代碼': stock_id,
                    '交易編號': trade_num,
                    '進場時間': trade.entry_time,
                    '進場價格': trade.entry_price,
                    '進場Ratio': trade.entry_ratio,
                    '進場外盤3秒(M)': trade.entry_outside_volume_3s / 1000000 if trade.entry_outside_volume_3s else 0,
                    '出場批次': len(trade.trailing_exit_details) + 1,
                    '出場類型': '最終清倉',
                    '出場時間': trade.final_exit_time,
                    '出場價格': trade.final_exit_price,
                    '出場比例': remaining_ratio * 100,
                    '出場原因': trade.final_exit_reason,
                    '實際出場均價': exit_price,
                    '總損益金額': pnl,
                    '總損益百分比': pnl_percent
                }
                trade_details.append(detail)

    def _add_two_stage_exit_details(self, trade_details: List[Dict], trade: TradeRecord, trade_num: int,
                                     stock_id: str, exit_price: float, pnl: float, pnl_percent: float) -> None:
        """添加兩階段出場詳情"""
        batch = 1

        # 第一階段出場（減碼50%）
        if trade.partial_exit_time:
            detail = {
                '股票代碼': stock_id,
                '交易編號': trade_num,
                '進場時間': trade.entry_time,
                '進場價格': trade.entry_price,
                '進場Ratio': trade.entry_ratio,
                '進場外盤3秒(M)': trade.entry_outside_volume_3s / 1000000 if trade.entry_outside_volume_3s else 0,
                '出場批次': batch,
                '出場類型': '減碼50%',
                '出場時間': trade.partial_exit_time,
                '出場價格': trade.partial_exit_price,
                '出場比例': 50.0,
                '出場原因': trade.partial_exit_reason,
                '實際出場均價': exit_price,
                '總損益金額': pnl,
                '總損益百分比': pnl_percent
            }
            trade_details.append(detail)
            batch += 1

        # 回補進場（如果有）
        if trade.reentry_time:
            detail = {
                '股票代碼': stock_id,
                '交易編號': trade_num,
                '進場時間': trade.entry_time,
                '進場價格': trade.entry_price,
                '進場Ratio': trade.entry_ratio,
                '進場外盤3秒(M)': trade.entry_outside_volume_3s / 1000000 if trade.entry_outside_volume_3s else 0,
                '出場批次': '回補',
                '出場類型': '回補進場',
                '出場時間': trade.reentry_time,
                '出場價格': trade.reentry_price,
                '出場比例': '-',
                '出場原因': '價格創新高且外盤增加',
                '實際出場均價': exit_price,
                '總損益金額': pnl,
                '總損益百分比': pnl_percent
            }
            trade_details.append(detail)

        # 最終出場（清倉）
        if trade.final_exit_time:
            detail = {
                '股票代碼': stock_id,
                '交易編號': trade_num,
                '進場時間': trade.entry_time,
                '進場價格': trade.entry_price,
                '進場Ratio': trade.entry_ratio,
                '進場外盤3秒(M)': trade.entry_outside_volume_3s / 1000000 if trade.entry_outside_volume_3s else 0,
                '出場批次': batch,
                '出場類型': '清倉' if not trade.reentry_time else '回補後清倉',
                '出場時間': trade.final_exit_time,
                '出場價格': trade.final_exit_price,
                '出場比例': 50.0 if trade.partial_exit_time else 100.0,
                '出場原因': trade.final_exit_reason,
                '實際出場均價': exit_price,
                '總損益金額': pnl,
                '總損益百分比': pnl_percent
            }
            trade_details.append(detail)

File: CSharp/BacktestModule/Analytics/TradeDetails.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BacktestModule.Core;
using BacktestModule.Strategy;

namespace BacktestModule.Analytics
{
    /// <summary>
    /// Trade details processor.
    /// Converts TradeRecord objects into flat detail dictionaries for CSV export.
    /// Each exit batch gets its own row.
    /// </summary>
    public class TradeDetailsProcessor
    {
        /// <summary>
        /// Collects detailed trade records for a single stock.
        /// Each exit batch is a separate dictionary entry for CSV row output.
        /// </summary>
        /// <param name="trades">List of trade records.</param>
        /// <param name="stockId">Stock code.</param>
        /// <param name="date">Date string.</param>
        /// <returns>List of detail dictionaries (one per exit batch).</returns>
        public List<Dictionary<string, object>> CollectTradeDetails(
            List<TradeRecord> trades, string stockId, string date)
        {
            var tradeDetails = new List<Dictionary<string, object>>();

            for (int i = 0; i < trades.Count; i++)
            {
                var trade = trades[i];
                int tradeNum = i + 1;

                // Calculate actual exit price (weighted average)
                double exitPrice = CalculateActualExitPrice(trade);
                double pnl = (exitPrice - trade.EntryPrice) * 1000; // 1 lot
                double pnlPercent = ((exitPrice - trade.EntryPrice) / trade.EntryPrice) * 100;

                // Handle trailing stop exits (may have multiple batches)
                if (trade.TrailingExitDetails != null && trade.TrailingExitDetails.Count > 0)
                {
                    AddTrailingExitDetails(tradeDetails, trade, tradeNum, stockId, exitPrice, pnl, pnlPercent);
                }
                // Handle two-stage exit mode
                else if (trade.PartialExitTime.HasValue || trade.FinalExitTime.HasValue)
                {
                    AddTwoStageExitDetails(tradeDetails, trade, tradeNum, stockId, exitPrice, pnl, pnlPercent);
                }
            }

            return tradeDetails;
        }

        /// <summary>
        /// Calculates the actual exit price as a weighted average across all exit batches.
        /// </summary>
        /// <param name="trade">The trade record.</param>
        /// <returns>Weighted average exit price.</returns>
        public double CalculateActualExitPrice(TradeRecord trade)
        {
            if (trade.TrailingExitDetails != null && trade.TrailingExitDetails.Count > 0)
            {
                // Trailing stop exit: calculate weighted average price
                double totalRatio = 0.0;
                double weightedPrice = 0.0;

                foreach (var exitDetail in trade.TrailingExitDetails)
                {
                    double price = exitDetail.ContainsKey("price") ? Convert.ToDouble(exitDetail["price"]) : 0.0;
                    double ratio = exitDetail.ContainsKey("ratio") ? Convert.ToDouble(exitDetail["ratio"]) : 0.0;
                    weightedPrice += price * ratio;
                    totalRatio += ratio;
                }

                // If there is remaining position (entry price protection or market close)
                if (trade.FinalExitPrice.HasValue && totalRatio < 1.0)
                {
                    double remainingRatio = 1.0 - totalRatio;
                    weightedPrice += trade.FinalExitPrice.Value * remainingRatio;
                    totalRatio = 1.0;
                }

                return totalRatio > 0 ? weightedPrice / totalRatio : trade.EntryPrice;
            }
            else if (trade.PartialExitPrice.HasValue && trade.FinalExitPrice.HasValue)
            {
                // Two-stage exit
                if (trade.ReentryPrice.HasValue)
                {
                    // With reentry: 50% at partial exit, 50% at final exit
                    return trade.PartialExitPrice.Value * 0.5 + trade.FinalExitPrice.Value * 0.5;
                }
                else
                {
                    // Without reentry: 50% at partial exit, 50% at final exit
                    return trade.PartialExitPrice.Value * 0.5 + trade.FinalExitPrice.Value * 0.5;
                }
            }
            else if (trade.PartialExitPrice.HasValue)
            {
                // Only partial exit
                return trade.PartialExitPrice.Value;
            }
            else if (trade.FinalExitPrice.HasValue)
            {
                // Only final exit
                return trade.FinalExitPrice.Value;
            }
            else
            {
                // No exit (should not happen in theory)
                return trade.EntryPrice;
            }
        }

        /// <summary>
        /// Adds trailing stop exit detail rows.
        /// </summary>
        private void AddTrailingExitDetails(
            List<Dictionary<string, object>> tradeDetails,
            TradeRecord trade, int tradeNum, string stockId,
            double exitPrice, double pnl, double pnlPercent)
        {
            // Each trailing stop exit batch
            for (int j = 0; j < trade.TrailingExitDetails.Count; j++)
            {
                var exitDetail = trade.TrailingExitDetails[j];
                string level = exitDetail.ContainsKey("level") ? exitDetail["level"]?.ToString() : "";
                double price = exitDetail.ContainsKey("price") ? Convert.ToDouble(exitDetail["price"]) : 0.0;
                double ratio = exitDetail.ContainsKey("ratio") ? Convert.ToDouble(exitDetail["ratio"]) : 0.0;
                object time = exitDetail.ContainsKey("time") ? exitDetail["time"] : null;

                double outsideVol3s = trade.EntryOutsideVolume3s > 0
                    ? trade.EntryOutsideVolume3s / 1_000_000.0
                    : 0.0;

                var detail = new Dictionary<string, object>
                {
                    ["\u80A1\u7968\u4EE3\u78BC"] = stockId,                                          // 股票代碼
                    ["\u4EA4\u6613\u7DE8\u865F"] = tradeNum,                                          // 交易編號
                    ["\u9032\u5834\u6642\u9593"] = trade.EntryTime,                                   // 進場時間
                    ["\u9032\u5834\u50F9\u683C"] = trade.EntryPrice,                                  // 進場價格
                    ["\u9032\u5834Ratio"] = trade.EntryRatio,                                         // 進場Ratio
                    ["\u9032\u5834\u5916\u76E43\u79D2(M)"] = outsideVol3s,                            // 進場外盤3秒(M)
                    ["\u51FA\u5834\u6279\u6B21"] = j + 1,                                             // 出場批次
                    ["\u51FA\u5834\u985E\u578B"] = $"\u79FB\u52D5\u505C\u5229_{level}",               // 出場類型: 移動停利_{level}
                    ["\u51FA\u5834\u6642\u9593"] = time,                                              // 出場時間
                    ["\u51FA\u5834\u50F9\u683C"] = price,                                             // 出場價格
                    ["\u51FA\u5834\u6BD4\u4F8B"] = ratio * 100,                                       // 出場比例
                    ["\u51FA\u5834\u539F\u56E0"] = $"\u8DCC\u7834{level}\u4F4E\u9EDE",                // 出場原因: 跌破{level}低點
                    ["\u5BE6\u969B\u51FA\u5834\u5747\u50F9"] = exitPrice,                             // 實際出場均價
                    ["\u7E3D\u640D\u76CA\u91D1\u984D"] = pnl,                                        // 總損益金額
                    ["\u7E3D\u640D\u76CA\u767E\u5206\u6BD4"] = pnlPercent                            // 總損益百分比
                };
                tradeDetails.Add(detail);
            }

            // If there is entry price protection or market close exit (remaining position)
            if (trade.FinalExitTime.HasValue && !string.IsNullOrEmpty(trade.FinalExitReason))
            {
                double remainingRatio = 1.0 - trade.TrailingExitDetails.Sum(
                    e => e.ContainsKey("ratio") ? Convert.ToDouble(e["ratio"]) : 0.0);

                if (remainingRatio > 0)
                {
                    double outsideVol3s = trade.EntryOutsideVolume3s > 0
                        ? trade.EntryOutsideVolume3s / 1_000_000.0
                        : 0.0;

                    var detail = new Dictionary<string, object>
                    {
                        ["\u80A1\u7968\u4EE3\u78BC"] = stockId,
                        ["\u4EA4\u6613\u7DE8\u865F"] = tradeNum,
                        ["\u9032\u5834\u6642\u9593"] = trade.EntryTime,
                        ["\u9032\u5834\u50F9\u683C"] = trade.EntryPrice,
                        ["\u9032\u5834Ratio"] = trade.EntryRatio,
                        ["\u9032\u5834\u5916\u76E43\u79D2(M)"] = outsideVol3s,
                        ["\u51FA\u5834\u6279\u6B21"] = trade.TrailingExitDetails.Count + 1,            // 出場批次
                        ["\u51FA\u5834\u985E\u578B"] = "\u6700\u7D42\u6E05\u5009",                    // 出場類型: 最終清倉
                        ["\u51FA\u5834\u6642\u9593"] = trade.FinalExitTime,
                        ["\u51FA\u5834\u50F9\u683C"] = trade.FinalExitPrice,
                        ["\u51FA\u5834\u6BD4\u4F8B"] = remainingRatio * 100,
                        ["\u51FA\u5834\u539F\u56E0"] = trade.FinalExitReason,
                        ["\u5BE6\u969B\u51FA\u5834\u5747\u50F9"] = exitPrice,
                        ["\u7E3D\u640D\u76CA\u91D1\u984D"] = pnl,
                        ["\u7E3D\u640D\u76CA\u767E\u5206\u6BD4"] = pnlPercent
                    };
                    tradeDetails.Add(detail);
                }
            }
        }

        /// <summary>
        /// Adds two-stage exit detail rows (partial exit + optional reentry + final exit).
        /// </summary>
        private void AddTwoStageExitDetails(
            List<Dictionary<string, object>> tradeDetails,
            TradeRecord trade, int tradeNum, string stockId,
            double exitPrice, double pnl, double pnlPercent)
        {
            int batch = 1;
            double outsideVol3s = trade.EntryOutsideVolume3s > 0
                ? trade.EntryOutsideVolume3s / 1_000_000.0
                : 0.0;

            // Stage 1: Partial exit (reduce 50%)
            if (trade.PartialExitTime.HasValue)
            {
                var detail = new Dictionary<string, object>
                {
                    ["\u80A1\u7968\u4EE3\u78BC"] = stockId,
                    ["\u4EA4\u6613\u7DE8\u865F"] = tradeNum,
                    ["\u9032\u5834\u6642\u9593"] = trade.EntryTime,
                    ["\u9032\u5834\u50F9\u683C"] = trade.EntryPrice,
                    ["\u9032\u5834Ratio"] = trade.EntryRatio,
                    ["\u9032\u5834\u5916\u76E43\u79D2(M)"] = outsideVol3s,
                    ["\u51FA\u5834\u6279\u6B21"] = batch,                                              // 出場批次
                    ["\u51FA\u5834\u985E\u578B"] = "\u6E1B\u78BC50%",                                  // 出場類型: 減碼50%
                    ["\u51FA\u5834\u6642\u9593"] = trade.PartialExitTime,
                    ["\u51FA\u5834\u50F9\u683C"] = trade.PartialExitPrice,
                    ["\u51FA\u5834\u6BD4\u4F8B"] = 50.0,
                    ["\u51FA\u5834\u539F\u56E0"] = trade.PartialExitReason ?? "",
                    ["\u5BE6\u969B\u51FA\u5834\u5747\u50F9"] = exitPrice,
                    ["\u7E3D\u640D\u76CA\u91D1\u984D"] = pnl,
                    ["\u7E3D\u640D\u76CA\u767E\u5206\u6BD4"] = pnlPercent
                };
                tradeDetails.Add(detail);
                batch++;
            }

            // Reentry (if any)
            if (trade.ReentryTime.HasValue)
            {
                var detail = new Dictionary<string, object>
                {
                    ["\u80A1\u7968\u4EE3\u78BC"] = stockId,
                    ["\u4EA4\u6613\u7DE8\u865F"] = tradeNum,
                    ["\u9032\u5834\u6642\u9593"] = trade.EntryTime,
                    ["\u9032\u5834\u50F9\u683C"] = trade.EntryPrice,
                    ["\u9032\u5834Ratio"] = trade.EntryRatio,
                    ["\u9032\u5834\u5916\u76E43\u79D2(M)"] = outsideVol3s,
                    ["\u51FA\u5834\u6279\u6B21"] = (object)"\u56DE\u88DC",                             // 出場批次: 回補
                    ["\u51FA\u5834\u985E\u578B"] = "\u56DE\u88DC\u9032\u5834",                         // 出場類型: 回補進場
                    ["\u51FA\u5834\u6642\u9593"] = trade.ReentryTime,
                    ["\u51FA\u5834\u50F9\u683C"] = trade.ReentryPrice,
                    ["\u51FA\u5834\u6BD4\u4F8B"] = (object)"-",
                    ["\u51FA\u5834\u539F\u56E0"] = "\u50F9\u683C\u5275\u65B0\u9AD8\u4E14\u5916\u76E4\u589E\u52A0",  // 價格創新高且外盤增加
                    ["\u5BE6\u969B\u51FA\u5834\u5747\u50F9"] = exitPrice,
                    ["\u7E3D\u640D\u76CA\u91D1\u984D"] = pnl,
                    ["\u7E3D\u640D\u76CA\u767E\u5206\u6BD4"] = pnlPercent
                };
                tradeDetails.Add(detail);
            }

            // Final exit (close all)
            if (trade.FinalExitTime.HasValue)
            {
                string exitType = trade.ReentryTime.HasValue
                    ? "\u56DE\u88DC\u5F8C\u6E05\u5009"    // 回補後清倉
                    : "\u6E05\u5009";                       // 清倉
                double exitRatio = trade.PartialExitTime.HasValue ? 50.0 : 100.0;

                var detail = new Dictionary<string, object>
                {
                    ["\u80A1\u7968\u4EE3\u78BC"] = stockId,
                    ["\u4EA4\u6613\u7DE8\u865F"] = tradeNum,
                    ["\u9032\u5834\u6642\u9593"] = trade.EntryTime,
                    ["\u9032\u5834\u50F9\u683C"] = trade.EntryPrice,
                    ["\u9032\u5834Ratio"] = trade.EntryRatio,
                    ["\u9032\u5834\u5916\u76E43\u79D2(M)"] = outsideVol3s,
                    ["\u51FA\u5834\u6279\u6B21"] = batch,
                    ["\u51FA\u5834\u985E\u578B"] = exitType,
                    ["\u51FA\u5834\u6642\u9593"] = trade.FinalExitTime,
                    ["\u51FA\u5834\u50F9\u683C"] = trade.FinalExitPrice,
                    ["\u51FA\u5834\u6BD4\u4F8B"] = exitRatio,
                    ["\u51FA\u5834\u539F\u56E0"] = trade.FinalExitReason ?? "",
                    ["\u5BE6\u969B\u51FA\u5834\u5747\u50F9"] = exitPrice,
                    ["\u7E3D\u640D\u76CA\u91D1\u984D"] = pnl,
                    ["\u7E3D\u640D\u76CA\u767E\u5206\u6BD4"] = pnlPercent
                };
                tradeDetails.Add(detail);
            }
        }
    }
}
